File: feb2/mar13_Rec11/nknights.java
package mar13;

public class nknights {

	static int counter=0;
	public static boolean isKnightSafe(int[][] board, int i, int j) {
		int n=board.length;
		if(i-2 >=0 && j-1 >=0 && board[i-2][j-1]!=0)
			return false;
		if(i-1 >=0 && j-2 >=0 && board[i-1][j-2]!=0)
			return false;
		if(i+1 <n && j-2 >=0 && board[i+1][j-2]!=0)
			return false;
		if(i+2 <n && j-1 >=0 && board[i+2][j-1]!=0)
			return false;
		if(i+2 <n && j+1 <n && board[i+2][j+1]!=0)
			return false;
		if(i+1 <n && j+2 <n && board[i+1][j+2]!=0)
			return false;
		if(i-1 >=0 && j+2 <n && board[i-1][j+2]!=0)
			return false;
		if(i-2 >=0 && j+1 <n && board[i-2][j+1]!=0)
			return false;
		return true;
	}
	public static void nkc_b(int[][] board, int kpsf, int bno) { //nknights combination using boxes on levels
		if (bno == board.length * board.length) {
			if (kpsf == board.length) {
				counter++;
				System.out.println("***********" + counter + "***********");
				for (int i = 0; i < board.length; i++) {
					for (int j = 0; j < board.length; j++) {
						System.out.print(board[i][j] + " ");
					}
					System.out.println();
				}
				System.out.println("***********" + counter + "***********");
			}
			return;
		}
		nkc_b(board, kpsf, bno + 1);// if the box says no to the queen
		int i = bno / board.length;
		int j = bno % board.length;
		if (board[i][j] == 0 && isKnightSafe(board, i, j)) {
			board[i][j] = 1;
			nkc_b(board, kpsf + 1, bno + 1);// if the box says yes to the queen
			board[i][j] = 0;
		}
	}
	public static void main(String[] args) {
		int[][] board=new int[3][3];
		nkc_b(board, 0, 0);

	}

}

File: june14/july12_Rec8/nknights.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

bool istheKnightSafe(vector<vector<bool>>& board,int i,int j)
{
    if(i-2>=0 && j-1>=0 && board[i-2][j-1])
        return false;

    else if(i-1>=0 && j-2>=0 && board[i-1][j-2])
        return false;

    else if(i+1<board.size() && j>=2 && board[i+1][j-2])
        return false;

    else if(i+2<board.size() && j>=1 && board[i+2][j-1])
        return false;

    else if(i+2<board.size() && j+1<board.size() && board[i+2][j+1])
        return false;

    else if(i+1<board.size() && j+2<board.size() && board[i+1][j+2])
        return false;

    else if(i>=1 && j+2<board.size() && board[i-1][j+2])
        return false;

    else if(i>=2 && j+1<board.size() && board[i-2][j+1])
        return false;
    else
        return true;
}
int count=0;
void nknights(vector<vector<bool>>& board,int bno,int kpsf,string asf)
//last queen was placed in this row
{
    if(bno == board.size()*board.size())
    {
        if(kpsf == board.size())
        {
            cout << asf << endl;
            count++;
        }
        return;
    }

    nknights(board,bno+1,kpsf,asf);
    int i=bno/board.size();
    int j=bno%board.size();
    if(istheKnightSafe(board,i,j)) // if(board[i][j] == false) vli condition optional h
    //kyuki b0 b1 ko call kr ra h or b1 b2 ko to mtlb kisi b box pe pehle se true ho hi ni skta 
    {
        board[i][j] = true;
        nknights(board,bno+1,kpsf+1,asf+to_string(i)+to_string(j)+"-");
        board[i][j] = false;
    }
}

int main(int argc,char** argv)
{   
    vector<vector<bool>> board (3,vector<bool>(3,false));
    nknights(board,0,0,"");
    cout << count << endl;
}